README.md
# event-store-mongodb-adapter

[![Build Status](https://travis-ci.org/prooph/event-store-mongodb-adapter.svg?branch=master)](https://travis-ci.org/prooph/event-store-mongodb-adapter)
[![Coverage Status](https://coveralls.io/repos/prooph/event-store-mongodb-adapter/badge.svg?branch=master&service=github)](https://coveralls.io/github/prooph/event-store-mongodb-adapter?branch=master)
[![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/prooph/improoph)

MongoDB Adapter for [ProophEventStore](https://github.com/prooph/event-store)

#CAUTION: The mongoDB adapter is **NOT** compatible With prooph/event-store v7. MongoDB has limited ACID support which is not compatible with the new features of prooph/event-store. MongoDB is a great choice for a read model database, but unfortunately it cannot be used as an event-store that requires transactions across documents (events). 
Codeliner has published a [Gist](https://gist.github.com/codeliner/14a8d98d53efafdd35e851d76e89cc94) that shows a custom implementation of a possible mongoDB event store supporting a subset of the v7 features in a very limited way. If you want to try to bring mongoDB back, please get in touch with us and we can discuss it! 

**Support for the adapter will end at 31 December 2017.**

Requirements
------------

- MongoDB >= 4.0
- MongoDB PHP Driver >= 1.5.2

Transactions
-------------

The transaction write concern for this adapter is `majority`.

The transaction read concern for this adapter is `snapshot`.

You can disable transactions for this adapter.

Considerations
--------------

This adapter does not use the MongoDB ObjectId for its primary key, instead a UUID (string) is used.

We recommend the AggregateStreamStrategy as the best strategy to use with this adapter.

Keep in mind that transaction safety works only for a replica set. Sharded cluster support is planned for MongoDB 4.2.
Therefore it's not safe to use this adapter in a sharded cluster environment, as MongoDB can't guarantee transaction safety.

Stream can not be reset if iteration was started due the MongoDB cursor.

CHANGELOG.md
# Change Log

## [v3.0.1](https://github.com/prooph/event-store-mongodb-adapter/tree/v3.0.1) (2018-08-26)
[Full Changelog](https://github.com/prooph/event-store-mongodb-adapter/compare/v3.0.0...v3.0.1)

**Fixed bugs:**

- MongoDB concurrency check [\#65](https://github.com/prooph/event-store-mongodb-adapter/pull/65) ([sandrokeil](https://github.com/sandrokeil))

## [v3.0.0](https://github.com/prooph/event-store-mongodb-adapter/tree/v3.0.0) (2018-08-16)
[Full Changelog](https://github.com/prooph/event-store-mongodb-adapter/compare/v2.4.1...v3.0.0)

**Implemented enhancements:**

- Use MongoDB 4.0 multi document transactions
- Requires MongoDB >= 4.0
- Requires MongoDB PHP Driver >= 1.5.2
- Creates collection indexes in background

## [v2.4.1](https://github.com/prooph/event-store-mongodb-adapter/tree/v2.4.1) (2017-02-21)
[Full Changelog](https://github.com/prooph/event-store-mongodb-adapter/compare/v2.4.0...v2.4.1)

**Implemented enhancements:**

- Add compound index for transaction\_id and expire\_at [\#61](https://github.com/prooph/event-store-mongodb-adapter/pull/61) ([sandrokeil](https://github.com/sandrokeil))

**Closed issues:**

- Do we need an index on transaction\_id ? [\#60](https://github.com/prooph/event-store-mongodb-adapter/issues/60)
- use native MongoDB data types whenever possible [\#53](https://github.com/prooph/event-store-mongodb-adapter/issues/53)

It is necessary to add this index manually to existing collection streams. Run this in your mongo shell:

db.[your stream collection].createIndex({"transaction_id" : 1, "expire_at" : 1})

## [v2.4.0](https://github.com/prooph/event-store-mongodb-adapter/tree/v2.4.0) (2016-07-22)
[Full Changelog](https://github.com/prooph/event-store-mongodb-adapter/compare/v2.3.0...v2.4.0)

**Fixed bugs:**

- Create missing index for replaying [\#59](https://github.com/prooph/event-store-mongodb-adapter/pull/59) ([prolic](https://github.com/prolic))

## [v2.3.0](https://github.com/prooph/event-store-mongodb-adapter/tree/v2.3.0) (2016-07-04)
[Full Changelog](https://github.com/prooph/event-store-mongodb-adapter/compare/v2.2.0...v2.3.0)

**Implemented enhancements:**

- allow php7 [\#57](https://github.com/prooph/event-store-mongodb-adapter/pull/57) ([prolic](https://github.com/prolic))

## [v2.2.0](https://github.com/prooph/event-store-mongodb-adapter/tree/v2.2.0) (2016-06-28)
[Full Changelog](https://github.com/prooph/event-store-mongodb-adapter/compare/v2.1.1...v2.2.0)

**Implemented enhancements:**

- Concurrency exception [\#56](https://github.com/prooph/event-store-mongodb-adapter/pull/56) ([prolic](https://github.com/prolic))

**Fixed bugs:**

- better concurrency handling [\#58](https://github.com/prooph/event-store-mongodb-adapter/pull/58) ([prolic](https://github.com/prolic))

## [v2.1.1](https://github.com/prooph/event-store-mongodb-adapter/tree/v2.1.1) (2016-05-14)
[Full Changelog](https://github.com/prooph/event-store-mongodb-adapter/compare/v2.1.0...v2.1.1)

## [v2.1.0](https://github.com/prooph/event-store-mongodb-adapter/tree/v2.1.0) (2016-05-08)
[Full Changelog](https://github.com/prooph/event-store-mongodb-adapter/compare/v2.0.3...v2.1.0)

**Merged pull requests:**

- Prepare 2.1.0 Release [\#55](https://github.com/prooph/event-store-mongodb-adapter/pull/55) ([codeliner](https://github.com/codeliner))
- Readme typo [\#54](https://github.com/prooph/event-store-mongodb-adapter/pull/54) ([jpkleemans](https://github.com/jpkleemans))

## [v2.0.3](https://github.com/prooph/event-store-mongodb-adapter/tree/v2.0.3) (2016-04-01)
[Full Changelog](https://github.com/prooph/event-store-mongodb-adapter/compare/v2.0.2...v2.0.3)

**Fixed bugs:**

- Unique index for aggregate\_id + version required to handle concurrent writes [\#49](https://github.com/prooph/event-store-mongodb-adapter/issues/49)

**Merged pull requests:**

- update factories to interop-config 1.0 [\#52](https://github.com/prooph/event-store-mongodb-adapter/pull/52) ([sandrokeil](https://github.com/sandrokeil))

## [v2.0.2](https://github.com/prooph/event-store-mongodb-adapter/tree/v2.0.2) (2016-03-04)
[Full Changelog](https://github.com/prooph/event-store-mongodb-adapter/compare/v2.0.1...v2.0.2)

**Implemented enhancements:**

- Update to coveralls ^1.0 [\#51](https://github.com/prooph/event-store-mongodb-adapter/pull/51) ([prolic](https://github.com/prolic))

**Fixed bugs:**

- add unique index for aggregate\_id + version [\#50](https://github.com/prooph/event-store-mongodb-adapter/pull/50) ([prolic](https://github.com/prolic))

**Closed issues:**

- Update to coveralls ^1.0 [\#47](https://github.com/prooph/event-store-mongodb-adapter/issues/47)

## [v2.0.1](https://github.com/prooph/event-store-mongodb-adapter/tree/v2.0.1) (2015-12-08)
[Full Changelog](https://github.com/prooph/event-store-mongodb-adapter/compare/v2.0...v2.0.1)

**Fixed bugs:**

- fix exception message typo [\#46](https://github.com/prooph/event-store-mongodb-adapter/pull/46) ([prolic](https://github.com/prolic))

## [v2.0](https://github.com/prooph/event-store-mongodb-adapter/tree/v2.0) (2015-11-22)
[Full Changelog](https://github.com/prooph/event-store-mongodb-adapter/compare/v2.0-beta.1...v2.0)

**Fixed bugs:**

- it can commit and rollback empty transactions [\#40](https://github.com/prooph/event-store-mongodb-adapter/pull/40) ([prolic](https://github.com/prolic))

**Merged pull requests:**

- v2.0 [\#45](https://github.com/prooph/event-store-mongodb-adapter/pull/45) ([codeliner](https://github.com/codeliner))
- fix typo [\#44](https://github.com/prooph/event-store-mongodb-adapter/pull/44) ([prolic](https://github.com/prolic))
- remove php7 support for now [\#43](https://github.com/prooph/event-store-mongodb-adapter/pull/43) ([prolic](https://github.com/prolic))
- Test replay of two aggregates [\#39](https://github.com/prooph/event-store-mongodb-adapter/pull/39) ([codeliner](https://github.com/codeliner))


src/MongoDBEventStoreAdapter.php
<?php
/*
 * This file is part of the prooph/event-store-mongodb-adapter.
 * (c) 2014 - 2015 prooph software GmbH <contact@prooph.de>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 *
 * Date: 08/06/15 - 21:58
 */

namespace Prooph\EventStore\Adapter\MongoDb;

use Assert\Assertion;
use DateTimeInterface;
use Iterator;
use MongoDB\BSON\UTCDatetime;
use MongoDB\Driver\BulkWrite;
use MongoDB\Driver\Manager;
use MongoDB\Driver\Query;
use MongoDB\Driver\ReadPreference;
use MongoDB\Driver\WriteConcern;
use MongoDB\Operation\CreateIndexes;
use Prooph\Common\Messaging\Message;
use Prooph\Common\Messaging\MessageConverter;
use Prooph\Common\Messaging\MessageFactory;
use Prooph\EventStore\Adapter\Adapter;
use Prooph\EventStore\Adapter\Feature\CanHandleTransaction;
use Prooph\EventStore\Exception\RuntimeException;
use Prooph\EventStore\Exception\StreamNotFoundException;
use Prooph\EventStore\Stream\Stream;
use Prooph\EventStore\Stream\StreamName;
use Rhumsaa\Uuid\Uuid;

/**
 * EventStore Adapter for Mongo DB
 *
 * Class MongoDBEventStoreAdapter
 * @package Prooph\EventStore\Adapter\MongoDb
 */
final class MongoDBEventStoreAdapter implements Adapter, CanHandleTransaction
{
    /**
     * @var MessageFactory
     */
    private $messageFactory;

    /**
     * @var MessageConverter
     */
    private $messageConverter;

    /**
     * @var Manager
     */
    private $manager;

    /**
     * @var string
     */
    private $dbName;

    /**
     * @var StreamName
     */
    private $currentStreamName;

    /**
     * The transaction id, if currently in transaction, otherwise null
     *
     * @var Uuid|null
     */
    private $transactionId;

    /**
     * @var WriteConcern|null
     */
    private $writeConcern;

    /**
     * Transaction timeout in seconds
     *
     * @var int
     */
    private $transactionTimeout = 50;

    /**
     * Custom sourceType to collection mapping
     *
     * @var array
     */
    private $streamCollectionMap = [];

    /**
     * @var ReadPreference
     */
    private $readPreference;

    /**
     * @param MessageFactory $messageFactory
     * @param MessageConverter $messageConverter
     * @param Manager $manager
     * @param string $dbName
     * @param WriteConcern|null $writeConcern
     * @param int|null $transactionTimeout
     * @param array $streamCollectionMap
     */
    public function __construct(
        MessageFactory $messageFactory,
        MessageConverter $messageConverter,
        Manager $manager,
        $dbName,
        WriteConcern $writeConcern = null,
        $transactionTimeout = null,
        array $streamCollectionMap = []
    ) {
        Assertion::minLength($dbName, 1, 'Mongo database name is missing');

        $this->messageFactory   = $messageFactory;
        $this->messageConverter = $messageConverter;
        $this->manager          = $manager;
        $this->dbName           = $dbName;
        $this->streamCollectionMap = $streamCollectionMap;
        $this->readPreference   = new ReadPreference(ReadPreference::RP_PRIMARY);

        if (null === $writeConcern) {
            $writeConcern = new WriteConcern(1, 0, true);
        }

        $this->writeConcern = $writeConcern;

        if (null !== $transactionTimeout) {
            Assertion::min($transactionTimeout, 1, 'Transaction timeout must be a positive integer');

            $this->transactionTimeout = $transactionTimeout;
        }
    }

    /**
     * @param Stream $stream
     * @throws RuntimeException If creation of stream fails
     * @return void
     */
    public function create(Stream $stream)
    {
        if (!$stream->streamEvents()->valid()) {
            throw new RuntimeException(
                sprintf(
                    "Cannot create empty stream %s. %s requires at least one event to extract metadata information",
                    $stream->streamName()->toString(),
                    __CLASS__
                )
            );
        }

        $streamName = $stream->streamName();

        $this->createIndexes($streamName);

        $this->appendTo($streamName, $stream->streamEvents());
    }

    /**
     * @param StreamName $streamName
     * @param Iterator $streamEvents
     * @throws StreamNotFoundException If stream does not exist
     * @return void
     */
    public function appendTo(StreamName $streamName, Iterator $streamEvents)
    {
        if ($this->currentStreamName !== null && $this->currentStreamName->toString() !== $streamName->toString()) {
            throw new \RuntimeException('Cannot write to different stream streams in one transaction');
        }

        $this->currentStreamName = $streamName;

        $bulk = new BulkWrite(['ordered' => false]);

        foreach ($streamEvents as $streamEvent) {
            $data = $this->prepareEventData($streamName, $streamEvent);
            $bulk->insert($data);
        }

        //var_dump($data);
        //try {

            $this->manager->executeBulkWrite(
                $this->dbName . '.' . $this->getCollectionName($this->currentStreamName),
                $bulk,
                $this->writeConcern
            );
//        } catch (\Exception $e) {
//            var_dump(get_class($e));
//            var_dump($e->getMessage());
//            die;
//        }
    }

    /**
     * @param StreamName $streamName
     * @param Message $e
     * @return array
     */
    private function prepareEventData(StreamName $streamName, Message $e)
    {
        $eventArr = $this->messageConverter->convertToArray($e);

        $eventData = [
            '_id'         => $eventArr['uuid'],
            'version'     => $eventArr['version'],
            'event_name'  => $eventArr['message_name'],
            'payload'     => $eventArr['payload'],
            'created_at'  => $eventArr['created_at']->format('Y-m-d\TH:i:s.u'),
        ];

        foreach ($eventArr['metadata'] as $key => $value) {
            $eventData[$key] = (string) $value;
        }

        if (null !== $this->transactionId) {
            $eventData['transaction_id'] = $this->transactionId->toString();

            $time = time() . '000';

            $eventData['expire_at'] = new UTCDatetime($time);
        }

        return $eventData;
    }

    /**
     * @param StreamName $streamName
     * @param null|int $minVersion
     * @return Stream|null
     */
    public function load(StreamName $streamName, $minVersion = null)
    {
        $events = $this->loadEvents($streamName, [], $minVersion);

        return new Stream($streamName, $events);
    }

    /**
     * @param StreamName $streamName
     * @param array $metadata
     * @param null|int $minVersion
     * @return MongoDBStreamIterator
     * @throws StreamNotFoundException
     */
    public function loadEvents(StreamName $streamName, array $metadata = [], $minVersion = null)
    {
        $query = $metadata;

        if (null !== $minVersion) {
            $query['version'] = ['$gte' => $minVersion];
        }

        $query['expire_at'] = ['$exists' => false];
        $query['transaction_id'] = ['$exists' => false];

        $query = new Query($query, [
            'sort' => [
                'version' => 1
            ]
        ]);

        $namespace = $this->dbName . '.' . $this->getCollectionName($streamName);

        return new MongoDBStreamIterator($this->manager, $namespace, $query, $this->readPreference, $this->messageFactory, $metadata);
    }

    /**
     * @param StreamName $streamName
     * @param DateTimeInterface|null $since
     * @param array $metadata
     * @return MongoDBStreamIterator
     */
    public function replay(StreamName $streamName, DateTimeInterface $since = null, array $metadata = [])
    {
        $query = $metadata;

        if (null !== $since) {
            $query['created_at'] = ['$gt' => $since->format('Y-m-d\TH:i:s.u')];
        }

        $query['expire_at'] = ['$exists' => false];
        $query['transaction_id'] = ['$exists' => false];

        $query = new Query($query, [
            'sort' => [
                'created_at' => 1,
                'version' => 1
            ]
        ]);

        $namespace = $this->dbName . '.' . $this->getCollectionName($streamName);

        return new MongoDBStreamIterator($this->manager, $namespace, $query, $this->readPreference, $this->messageFactory, $metadata);
    }

    /**
     * Begin transaction
     *
     * @return void
     */
    public function beginTransaction()
    {
        if (null !== $this->transactionId) {
            throw new \RuntimeException('Transaction already started');
        }

        $this->transactionId = Uuid::uuid4();
    }

    /**
     * Commit transaction
     *
     * @return void
     */
    public function commit()
    {
        if (! $this->currentStreamName) {
            $this->transactionId = null;
            return;
        }

        $bulk = new BulkWrite(['ordered' => false]);
        $bulk->update(
            [
                'transaction_id' => $this->transactionId->toString(),
                '$isolated' => 1,
            ],
            [
                '$unset' => [
                    'expire_at' => 1,
                    'transaction_id' => 1
                ]
            ],
            [
                'multi' => true
            ]
        );

        $this->manager->executeBulkWrite(
            $this->dbName . '.' . $this->getCollectionName($this->currentStreamName),
            $bulk,
            $this->writeConcern
        );

        $this->transactionId = null;
        $this->currentStreamName = null;
    }

    /**
     * Rollback transaction
     *
     * @return void
     */
    public function rollback()
    {
        if (! $this->currentStreamName) {
            $this->transactionId = null;
            return;
        }

        $bulk = new BulkWrite(['ordered' => false]);
        $bulk->delete(
            [
                'transaction_id' => $this->transactionId
            ],
            [
                'limit' => 0
            ]
        );

        $this->manager->executeBulkWrite(
            $this->dbName . '.' . $this->getCollectionName($this->currentStreamName),
            $bulk,
            $this->writeConcern
        );
    }

    /**
     * @param StreamName $streamName
     * @return void
     */
    private function createIndexes(StreamName $streamName)
    {
        $createIndexes = new CreateIndexes($this->dbName, $this->getCollectionName($streamName), [
            [
                'key' => [
                    '_id' => 1,
                    'transaction_id' => 1,
                ],
                'unique' => true,
                'background' => true,
            ],
            [
                'key' => [
                    'expire_at' => 1,
                ],
                'expireAfterSeconds' => $this->transactionTimeout,
            ]
        ]);

        $createIndexes->execute($this->manager->selectServer($this->readPreference));
    }

    /**
     * @param StreamName $streamName
     * @return string
     */
    private function getCollectionName(StreamName $streamName)
    {
        if (isset($this->streamCollectionMap[$streamName->toString()])) {
            $collectionName = $this->streamCollectionMap[$streamName->toString()];
        } else {
            $collectionName = strtolower($this->getShortStreamName($streamName));
            if (strpos($collectionName, "_stream") === false) {
                $collectionName.= "_stream";
            }
        }

        return $collectionName;
    }

    /**
     * @param StreamName $streamName
     * @return string
     */
    private function getShortStreamName(StreamName $streamName)
    {
        $streamName = str_replace('-', '_', $streamName->toString());
        return implode('', array_slice(explode('\\', $streamName), -1));
    }
}
